### ISSUE TITLE
Items with same root item spread over multiple config files

### ISSUE THREAD

**Issue by mayrjohannes** [smarthomeNG/smarthome#180]
I do have the following config files:
licht.conf:
[global]
   [[counter_licht]]
    ...
[eg]
    ...

beschattung.conf
[global]
   [[counter_besch]]
    ...
[eg]
    ...

So far this was never a problem. But now i realised that this can cause troubles for eval.
e.g.
[global]
     [[counter_besch]]
      type = num
        ...
   [[counter_besch_bool]]
       type = bool
       eval = 1 if sh.global.counter_besch() > 0 else 0
       eval.trigger = global.counter_besch

leads to the following warning (i use different item names in my real config):
2016-12-13  20:09:51 WARNING  global.counter.raffstore.alle Item global.counter.raffstore.alle: problem evaluating sh.global.counter.raffstore.eg() + sh.global.counter.raffstore.og(): invalid syntax (<string>, line 1)

renaming global to global2 in the beschattung.conf resolves that issue. But this degenerates my items structure.

Is this a wrong configuration from my side or a bug?

**Comment by psilo909** [comment]
where are u doing "sh.global.counter.raffstore.eg() + sh.global.counter.raffstore.og()"

**Comment by mayrjohannes** [comment]
The command is used in one of my config files. Adopting my examples I have config files like:
**licht.conf:**
[**global**]
[[counter_licht]]
...
[eg]
...

**beschattung.conf**
[**global**]
[[counter_besch]]
type = num
...
[[counter_besch2]]
type = num
...
[[counter_besch_bool]]
type = bool
eval = sh.global.counter_besch()  + sh.global.counter_besch2()
eval.trigger = global.counter_besch | global.counter_besch2
...
[eg]
...


replacing the **beschattung.conf** with
[**global# 2**]
[[counter_besch]]
type = num
...
[[counter_besch2]]
type = num
...
[[counter_besch_bool]]
type = bool
eval = sh.global.counter_besch()  + sh.global.counter_besch2()
eval.trigger = global.counter_besch | global.counter_besch2
...
[eg]
...

**Comment by ohinckel** [comment]
The problem is in the parsing logic which handles each file separately: the [parse](https://github.com/smarthomeNG/smarthome/blob/develop/lib/config.py#L21) method keeps track of  parents in a list for each level of the configuration file. Reading additional files the previously parents list gets lost and is initialized empty.

AFAIS, this should also be a problem with the items itself: they should completly be ignored. So, does a `ls` list all the items or are there some missing too?

**Comment by ohinckel** [comment]
I tried to verify this problem but I can't. My setup looks like this:
#### test.conf
```
[temp]
  [[level2]]
    [[[out]]]
      type = num
```
#### overwrite.conf
```
[temp]
  [[level2]]
    [[[overwrite]]]
      type = num
      eval = sh.temp.level2.out()
      eval_trigger = temp.level2.out
```
Updating (in my setup via the CLI plugin) the item `temp.level2.out` (declared in `test.conf`) updated the item `temp.level2.overwrite` (declared in `overwrite.conf`) without any problem:
```
2017-05-05 20:56:01 DEBUG    item         Main         Item temp.level2.out = 1 via CLI 127.0.0.1:39474 None -- item.py:__update:658
2017-05-05 20:56:01 DEBUG    scheduler    Main         Triggering temp.level2.overwrite - by: CLI source: 127.0.0.1:39474 dest: None value: {'value': 1, 'source': 'temp.level2.out' -- scheduler.py:trigger:168
2017-05-05 20:56:01 DEBUG    item         temp.level2.overwrite Item temp.level2.overwrite = 1 via Eval temp.level2.out None -- item.py:__update:658
```

**Comment by mayrjohannes** [comment]
Two weeks ago I figured out what was the problem. At leat I think.

My main group (in your example you used temp) was global. As long as global
is not used in any eval everything is fine, but using it in an eval I think
it gets mixed up with a keyword from Python?

I renamed all my global to glob and now its working without problems.

Best  Regards

Johannes Mayr

2017-05-05 21:02 GMT+02:00 ohinckel <notifications@github.com>:

**Comment by ohinckel** [comment]
Maybe the name "global" was used as plugin instance name too? Both items and plugins are registered as items in `sh` object (like the `sh.global.some.item()` or `sh.plugin.method()`). So one need to take case not overwriting stuff from plugin and item configuration. IMHO this is not a good implementation. I would prefer to use `sh.plugins.name.method()` or `sh.items.some.item()` instead.

I did not test with the `global` name to find out the problem (will do sometime) but the error indicates a syntax problem. So the expression can not be evaluated. Maybe the [parsing](https://github.com/smarthomeNG/smarthome/blob/73113e2478c5b538f772030b8daed94b0fea5f67/lib/item.py#L329) (and [this](https://github.com/smarthomeNG/smarthome/blob/73113e2478c5b538f772030b8daed94b0fea5f67/lib/item.py#L346)) is messing up things here?

**Comment by ohinckel** [comment]
Using `global` as root item generates indeed an error:
```
2017-05-07 19:10:51 WARNING  item         global.level2.overwrite Item global.level2.overwrite: problem evaluating sh.global.level2.out(): invalid syntax (<string>, line 1) -- item.py:__run_eval:626
```

It seems global is a keyword and can not be used in expressions:
```
>>> import collections
>>> a = collections.OrderedDict()
>>> vars(a)['other'] = 1
>>> vars(a)['global'] = 2
>>> a.other
1
>>> a.global
  File "<stdin>", line 1
    a.global
           ^
SyntaxError: invalid syntax
>>> vars(a)['and'] = 3
>>> a.and
  File "<stdin>", line 1
    a.and
        ^
SyntaxError: invalid syntax
>>> keyword.iskeyword('global')
True
>>> keyword.iskeyword('and')
True
>>> 
```

The solution would be not using keywords in item names at all. I created #207 for this issue which will ignore reserved Python keywords in configuration files (like other ignored things, digits or `set`/`get`, too).

**Comment by ohinckel** [comment]
I'll close this issue, since it was an error using Python keywords in configuration files (since they are mapped to dictionaries this will generate an error - see previous comment).

Using keywords is not allowed anymore (fixed with #207, which is merged already).
